### PR TITLE
Some changes in UnsafeBufferPointer to improve performance

### PR DESCRIPTION
**Pull request by eeckstein**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

**Comment by eeckstein** [comment]
@gribozavr, can you please review this?

**Comment by swiftix** [comment]
👍

**Comment by gribozavr** [comment]
@swift-ci Please test

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb [hunk 1]
```diff
@@ -57,6 +57,83 @@ public struct Unsafe${Mutable}BufferPointer<Element>
     return count
   }
 
+  public func index(after i: Int) -> Int {
+    // NOTE: this is a manual specialization of index movement for a Strideable
+    // index that is required for Array performance.  The optimizer is not
```
Comment (gribozavr) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: s/Array/UnsafeBufferPointer/

**Comment by gribozavr** [comment]
LGTM modulo the comment.

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please test

**Pull request merged**